pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,187 @@
+	}
+	conditionSeedRegistered = gardencorev1beta1helper.UpdatedCondition(conditionSeedRegistered, gardencorev1beta1.ConditionTrue, "SeedRegistered",
+		fmt.Sprintf("Shoot %s registered as seed", kutil.ObjectName(shoot)))
+
+	// Return success result
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) delete(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, logger *logrus.Entry) (reconcile.Result, error) {
+	// Check gardener finalizer
+	if !controllerutil.ContainsFinalizer(managedSeed, gardencorev1beta1.GardenerName) {
+		logger.Debug("Skipping ManagedSeed as it does not have a finalizer")
+		return reconcile.Result{}, nil
+	}
+
+	conditionShootExists := gardencorev1beta1helper.GetOrInitCondition(managedSeed.Status.Conditions, seedmanagementv1alpha1.ManagedSeedShootExists)
+	conditionSeedRegistered := gardencorev1beta1helper.GetOrInitCondition(managedSeed.Status.Conditions, seedmanagementv1alpha1.ManagedSeedSeedRegistered)
+
+	defer func() {
+		if err := updateStatus(ctx, r.gardenClient.Client(), managedSeed, conditionShootExists, conditionSeedRegistered); err != nil {
+			logger.Errorf("Could not update status: %+v", err)
+		}
+	}()
+
+	// Ensure shoot exists
+	shoot, err := r.ensureShootExists(ctx, managedSeed, &conditionShootExists, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile deletion
+	if err := r.actuator.Delete(ctx, managedSeed, shoot); err != nil {
+		message := fmt.Sprintf("Could not unregister seed: %+v", err)
+		conditionSeedRegistered = gardencorev1beta1helper.UpdatedCondition(conditionSeedRegistered, gardencorev1beta1.ConditionFalse, "SeedUnregistrationFailed", message)
+		logger.Error(message)
+		r.recorder.Eventf(managedSeed, corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, "%s", message)
+		return reconcile.Result{}, fmt.Errorf("could not unregister seed: %w", err)
+	}
+	conditionSeedRegistered = gardencorev1beta1helper.UpdatedCondition(conditionSeedRegistered, gardencorev1beta1.ConditionFalse, "SeedUnregistered",
+		fmt.Sprintf("Shoot %s unregistered as seed", kutil.ObjectName(shoot)))
+
+	// Return success result and remove finalizer
+	return reconcile.Result{}, controllerutils.PatchRemoveFinalizers(ctx, r.gardenClient.Client(), managedSeed, gardencorev1beta1.GardenerName)
+}
+
+func (r *reconciler) ensureShootExists(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, conditionShootExists *gardencorev1beta1.Condition, logger *logrus.Entry) (*gardencorev1beta1.Shoot, error) {
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := r.gardenClient.DirectClient().Get(ctx, kutil.Key(managedSeed.Namespace, managedSeed.Spec.Shoot.Name), shoot); err != nil {
+		if apierrors.IsNotFound(err) {
+			message := fmt.Sprintf("Shoot %s/%s not found", managedSeed.Namespace, managedSeed.Spec.Shoot.Name)
+			*conditionShootExists = gardencorev1beta1helper.UpdatedCondition(*conditionShootExists, gardencorev1beta1.ConditionFalse, "ShootNotFound", message)
```

Comment:
OK

Reply:
Actually, after taking a closer look and trying to do it, I found some issues with this approach.

* You may be right about the error case. But what about if the check succeeds, what status should the condition receive in this case? "true" would be wrong, as well as "false", so we would have to settle for "unknown" or "progressing", which is awkward.
* What about deletion? The "exists" check has to be performed there as well, but not the "reconciled" check. Again, on success, we would have to update the "reconciled" condition with "unknown" or "progressing", and it will never become "true", since we don't even perform this check.

I think it's much cleaner to have a separate condition for "exists" that can be updated independently of "reconciled", resulting in more straightforward controller logic and better user experience.

Let's please keep the condition unless you have really strong arguments to remove it.

Reply:
Ok, let me try to clarify what I meant.
I didn't mean to add the message/status of the current `ShootExists` condition to `ShootReconciled` in any case.
I was rather proposing to only add an error message indicating that the shoot doesn't exist in the error case.

Anyways, as this is a minor thing, it's fine for me to proceed with this for now and remove it later on, if we reach an agreement for removing it.

Reply:
From my perspective it's both more straightforward, more consistent, and less confusing for the user when checking something, to update its condition with `True` if the check succeeds and with `False` otherwise. The logic in the reconciler is currently following this principle. I don't think having one additional condition to be able to do this hurts in any way, quite the opposite. It could always be removed and the logic adapted somehow to deal with its absence, but I still don't see why would we want to do it.

Reply:
In any case, since you agree to keep it, let's resolve this now.